### PR TITLE
S3 packages

### PR DESCRIPTION
**Pull request by Bjvanminnen**
This is a semi-scary change, because it changes our CI process some. I've managed to have most of the logic live it's own module that is reasonably testable, but how it integrates with the entire system has not been significantly tested. There will probably be bugs to start. For now I've only done the work for code-studio, but it should be pretty easy to replicate this for apps once we're convinced it's working as expected.

Long explanation is at https://docs.google.com/document/d/1WlOrFUw1xXGjjcHWvGH3qbsv5m6xJ0vMnJu5rY2fK5w/edit#bookmark=id.w103gcxpcika

Shorter explanation is that rather than storing built packages in our git repo, we push them up to s3 keyed off of their commit hash. As part of the CI system, we then first attempt to pull down packages from s3 before potentially rebuilding.

Testing
- Ran a custom build-code-studio task, that ran only CODE_STUDIO_TASK (after commenting out the line that throws if you're not on staging/test), validated we create/upload package
- Made/built changes to code-studio in a separate branch and committed them. Ran the task again, and validated that we create/upload a new package
- Ran `bin/dashboard-server` and validated that as I moved between commits, it background updated the code-studio package by downloading from S3
- ran websites task

Note: When looking at the diff, there's a bunch of files being removed in the middle, and then some interesting files at the end (i.e DO scroll all the way down.)

### CODE REVIEW

File: Rakefile [hunk 1]
```diff
@@ -237,6 +238,14 @@ def install_npm
   end
 end
 
+def ensure_code_studio_package
+  # never download if we build our own
+  return if CDO.use_my_code_studio
+
+  packager = S3Packaging.new('code-studio', code_studio_dir, dashboard_dir('public/code-studio-package'))
+  raise "No valid package found" unless packager.update_from_s3
```
Comment (Bjvanminnen) [inline Rakefile]: Note: This means if you don't use_my_code_studio, and sync to a point before we started uploading to s3 or to a point where we don't have a package for some reason, this will fail. Not sure if there's a way we can be smarter about this.

File: lib/cdo/aws/s3_packaging.rb [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+    return nil unless File.exist?(filename)
+    IO.read(filename)
+  end
+
+  # Creates a zipped package of the provided assets folder
+  # @param sub_path [String] Path to built assets, relative to source_location
+  # @return tempfile object of package
+  private def create_package(sub_path)
+    package = Tempfile.new(@commit_hash)
+    # TODO - better recommendation for logging that puts?
```
Comment (Bjvanminnen) [inline lib/cdo/aws/s3_packaging.rb]: s/that/than Not sure what the recommendation is here. HipChat.log is probably more verbose than we want there?

File: lib/cdo/aws/s3_packaging.rb [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+    rescue Exception => e
+      puts "update_from_s3 failed: #{e.message}"
+      return false
+    end
+    return true
+  end
+
+  # creates a package from the given assets location and upload it to s3
+  # @return tempfile object of package
+  def upload_packge_to_s3(sub_path)
```
Comment (wjordan) [inline lib/cdo/aws/s3_packaging.rb]: typo in `packge`, or intentional (since it's referenced consistently throughout)?

Comment (Bjvanminnen) [inline lib/cdo/aws/s3_packaging.rb]: haha. not intentional, but i did do a global search/replace at one point :-P

File: Rakefile (hunk 1 above)
Comment (islemaster) [inline Rakefile]: If you sync to a point before we started uploading to s3, then this code won't exist to fail, right?

File: Rakefile [hunk 4]
```diff
@@ -277,6 +286,7 @@ namespace :install do
         code_studio_build = CDO.use_my_code_studio ? code_studio_dir('build') : 'code-studio-package'
         RakeUtils.ln_s code_studio_build, dashboard_dir('public','code-studio')
       end
+      ensure_code_studio_package
```
Comment (islemaster) [inline Rakefile:307]: Ensuring the code-studio package feels more like an install step for dashboard, not for code-studio itself.  Maybe even a build step for dashboard.

File: Rakefile (hunk 1 above)
Comment (philbogle) [inline Rakefile]: Style nit: I don't think a function call with side effects should be in a postfix unless clause.

I would find this more readable across two like this like this:
package_found = package.update_from_s3
raise "No valid package found" unless package_found

File: aws/build.rake [hunk 5]
```diff
@@ -94,7 +95,7 @@ def threaded_each(array, thread_count=2)
   threads.each(&:join)
 end
 
-if (rack_env?(:staging) && CDO.name == 'staging') || rack_env?(:development)
+if rack_env?(:staging) || rack_env?(:development)
```
Comment (philbogle) [inline aws/build.rake]: Why did this condition change?

File: aws/build.rake [hunk 6]
```diff
@@ -166,45 +154,32 @@ if (rack_env?(:staging) && CDO.name == 'staging') || rack_env?(:development)
...
-      code_studio_git_status = `git status --porcelain .`
-      code_studio_changed = !code_studio_git_status.strip.empty?
-      HipChat.log "<b>code-studio</b> package changes:\n#{code_studio_git_status}" if code_studio_changed
-    end
+#
+# Define the CODE STUDIO BUILD task.
+#
+CODE_STUDIO_NODE_MODULES = Dir.glob(code_studio_dir('node_modules', '**/*'))
+CODE_STUDIO_BUILD_PRODUCTS = [code_studio_dir('npm-debug.log')] + Dir.glob(code_studio_dir('build', '**/*'))
+CODE_STUDIO_SOURCE_FILES = Dir.glob(code_studio_dir('**/*')) - CODE_STUDIO_NODE_MODULES - CODE_STUDIO_BUILD_PRODUCTS
```
Comment (islemaster) [inline aws/build.rake]: Is this code to build up the list of source files still needed?

File: aws/build.rake (hunk 5 above)
Comment (Bjvanminnen) [inline aws/build.rake]: I don't think it changed. I think it's been a redundant check for some time, and when I saw it I was confused so I fixed it. Could change this separately if we want to reduce the change surface in this PR.

Comment (philbogle) [inline aws/build.rake]: Yes, if it's not required, it seems like this optimization should be in a different PR to reduce the risk and change surface.    

It seems possible that CDO.name != rack_env, at least in principle.

File: aws/build.rake (hunk 6 above)
Comment (Bjvanminnen) [inline aws/build.rake]: You could probably make the argument that it's not.

Right now, the code inside this block will not even get run if all that changed was node_modules (for example). In a world where we remove this list, we would check for a commit_hash file, compare the value there to git_latest_commit_hash, and if all that changed was node_modules, but there were no new commits, we would be done.

So yeah, I think we can remove these and make this simpler.

File: aws/build.rake [hunk 7]
```diff
@@ -166,45 +154,32 @@ if (rack_env?(:staging) && CDO.name == 'staging') || rack_env?(:development)
...
-        RakeUtils.system 'git', 'commit', '-m', Shellwords.escape(message)
-        RakeUtils.git_push
-        RakeUtils.system 'rm', '-f', deploy_dir('rebuild-code-studio')
-      end
-    else
-      HipChat.log '<b>code-studio</b> package unmodified, nothing to commit.'
-      RakeUtils.system 'rm', '-f', deploy_dir('rebuild-code-studio')
-    end
-  end
+  next if packager.update_from_s3
```
Comment (philbogle) [inline aws/build.rake]: "next" is confusing to me here; this isn't a loop body is it?

File: lib/cdo/aws/s3_packaging.rb [hunk 8]
```diff
@@ -0,0 +1,143 @@
...
+    package = create_package(sub_path)
+    raise "Generated different package for same contents" unless package_matches_download(package)
+    upload_package(package)
+    package
+  end
+
+  # Unzips package into target location
+  def decompress_package(package)
+    puts "Decompressing #{package.path}\nto #{@target_location}"
```
Comment (philbogle) [inline lib/cdo/aws/s3_packaging.rb]: Consider using Rails.logger instead of puts.

File: lib/cdo/aws/s3_packaging.rb [hunk 9]
```diff
@@ -0,0 +1,143 @@
...
+    FileUtils.mkdir_p(@target_location)
+    Dir.chdir(@target_location) do
+      # Clear out existing package
+      FileUtils.rm_rf Dir.glob("#{@target_location}/*")
```
Comment (philbogle) [inline lib/cdo/aws/s3_packaging.rb:57]: "rm -rf *" seems very dangerous in the event of a misconfiguration; e.g. suppose target_location was set to "/" or "".  

Could we remove just the existing package file?

**Comment by islemaster** [comment]
This is sweet.  Also somewhat complicated. I'm tempted to request a new file in docs/ or a wiki page (or even just a file-level comment in s3_packaging.rb) dedicated to explaining what's going on here.

File: lib/cdo/aws/s3_packaging.rb [hunk 10]
```diff
@@ -0,0 +1,143 @@
+require 'aws-sdk'
+
+class S3Packaging
+  BUCKET_NAME = 'cdo-build-package'
+
+  def initialize(package_name, source_location, target_location)
+    @client = Aws::S3::Client.new
```
Comment (philbogle) [inline lib/cdo/aws/s3_packaging.rb:19]: We should raise an exception in the case where any of the provided arguments are nil.

File: aws/build.rake (hunk 7 above)
Comment (Bjvanminnen) [inline aws/build.rake]: What's the right way to break out of a do block like this in ruby? I wasn't sure, did some googling, and next seemed to work...

File: lib/cdo/aws/s3_packaging.rb (hunk 2 above)
Comment (philbogle) [inline lib/cdo/aws/s3_packaging.rb]: The Ruby logger will prepend a timestamp and log level indicate (I for info, W for warn, etc) which makes it easier to grep and interpret the logged items.

File: lib/cdo/aws/s3_packaging.rb (hunk 9 above)
Comment (Bjvanminnen) [inline lib/cdo/aws/s3_packaging.rb:57]: target_location has the uncompressed package, so it is multiple files. we could just compress over top, but that means if commit A has a file, and commit B removes that file, it won't actually go away.

File: lib/cdo/aws/s3_packaging.rb (hunk 2 above)
Comment (philbogle) [inline lib/cdo/aws/s3_packaging.rb]: You can create a new logger pointed at STDOUT as follows.  This will be identical to puts except for the standard header.

require 'logger'
logger = Logger.new(STDOUT)
logger.info('this is a test')

File: lib/cdo/aws/s3_packaging.rb [hunk 11]
```diff
@@ -0,0 +1,143 @@
...
+  # @param sub_path [String] Path to built assets, relative to source_location
+  # @return tempfile object of package
+  private def create_package(sub_path)
+    package = Tempfile.new(@commit_hash)
+    # TODO - better recommendation for logging that puts?
+    puts "Creating #{package.path}"
+    Dir.chdir(@source_location + '/' + sub_path) do
+      # add a commit_hash file whose contents represent the key for this package
+      IO.write('commit_hash', @commit_hash)
+      # TODO - should i be using RakeUtils.system instead in these places?
```
Comment (philbogle) [inline lib/cdo/aws/s3_packaging.rb]: Using Rakeutils.system seems like a good idea for standardized behavior; e.g. it outputs the command being run.

File: lib/cdo/aws/s3_packaging.rb [hunk 12]
```diff
@@ -0,0 +1,143 @@
...
+    end
+    puts "Decompressed"
+  end
+
+  private def s3_key
+    "#{@package_name}/#{@commit_hash}.tar.gz"
+  end
+
+  # The hash of the package at target_location (or nil if there is not one)
+  private def target_commit_hash
```
Comment (philbogle) [inline lib/cdo/aws/s3_packaging.rb]: I think this would be easier to understand if you made the location parameter explicit; e.g. the caller would say "commit_hash(@target_location)".   That way they don't have to guess that target implies target location.

File: lib/cdo/aws/s3_packaging.rb [hunk 13]
```diff
@@ -0,0 +1,143 @@
+require 'aws-sdk'
+
+class S3Packaging
```
Comment (philbogle) [inline lib/cdo/aws/s3_packaging.rb:11]: Please a class comment that explains what this class does.

File: lib/cdo/aws/s3_packaging.rb [hunk 14]
```diff
@@ -0,0 +1,143 @@
...
+  BUCKET_NAME = 'cdo-build-package'
+
+  def initialize(package_name, source_location, target_location)
```
Comment (philbogle) [inline lib/cdo/aws/s3_packaging.rb:17]: Please add a comment explaining what source_location and target_location mean.

File: lib/cdo/aws/s3_packaging.rb [hunk 15]
```diff
@@ -0,0 +1,143 @@
...
+      `tar -zxf #{package1.path} -C #{dir1}`
+      Dir.mktmpdir do |dir2|
+        `tar -zxf #{package2.path} -C #{dir2}`
+        `diff -rq #{dir1} #{dir2}`
+      end
+    end
+    diff.empty?
+  end
+
+  # Downloads package from S3, throws error if given package doesnt exist on s3
```
Comment (philbogle) [inline lib/cdo/aws/s3_packaging.rb]: s/doesnt/doesn't/

**Comment by Bjvanminnen** [comment]
Code review feedback has been addressed. Feel free to take another look
